### PR TITLE
AO3-6013 Record date of wrangler's last wrangling activity

### PR DESCRIPTION
**Pull request by brianjaustin**
# Pull Request Checklist

* [X] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [X] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [X] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [X] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [X] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6013

## Purpose

Record the time in a new table whenever a tag wrangler
* Comments on a tag (including edits and deletes)
* Modifies a tag
* Modifies a tag relationship (common tagging)
* Modifies a meta-tagging

The timestamp is visible on the wrangler's profile for the wrangler and admins. For example:
new message with assigned fandoms

new message with no assigned fandoms

## Testing Instructions

Listed in the Jira ticket's description

## References

Ticket unblocked by this change: https://otwarchive.atlassian.net/browse/AO3-6028

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?
**Brian Austin (they/he)**

### CODE REVIEW

**Review by sarken — COMMENTED** [review]
Thank you for working on this! Tag Wrangling will really appreciate it.

I have two small comments for now and a topic for discussion for anyone who happens to be reading this: I'm not sure if doing this in the controller is the right approach. Should it be done in the models instead?

**Comment by jbeal-work** [comment]
I believe this should be done in the models.

**Comment by zz9pzza** [comment]
I believe this should be done in the models.

**Comment by brianjaustin** [comment]
> I believe this should be done in the models.

I initially went with the controller because it involves `current_user`, but I found `User.current_user` so the model makes sense to me

File: app/controllers/comments_controller.rb
Comment (brianjaustin) [reply app/controllers/comments_controller.rb]: Neat! Updated to reflect this as part of switching to doing this in the model

File: app/models/roles_user.rb
Comment (brianjaustin) [reply app/models/roles_user.rb]: No particular reason, I'm just getting used to Rails and that was what came to mind first. `destroy` looks clearer to me though so I've changed to use that instead

File: app/models/concerns/wranglable.rb
Comment (sarken) [inline app/models/concerns/wranglable.rb:13]: Could this be just `user: current_user`, since we already defined the variable on line 10?

File: app/models/comment.rb
Comment (sarken) [inline app/models/comment.rb:183]: Would it be at all possible to pull this out into a separate `after_create` callback? This existing one is... quite long and unhelpfully named, so it would be great if we could avoid making the problem worse by having `after_create :update_last_wrangling_activity` specifically for wrangling activity.

**Review by sarken — COMMENTED** [review]
This looks so much better in the model -- thanks! 

I just did a quick skim and left a coupe of comments; it's fine if you'd rather wait for a second, more thorough review before addressing any of them.

**Review by tickinginstant — COMMENTED** [review]
This is not a thorough review, but I wanted to flag an issue with using model callbacks for this, and I noticed a couple other things while I was at it.

I'd be inclined to take a hybrid model/controller approach, with an `around_action` hook in wrangling-related controllers to set a class variable like `User.current_user`. But there might also be a purely model-based approach -- maybe you could flag any tags being created/modified through `Taggable`? IDK.

File: app/models/concerns/wranglable.rb
Comment (sarken) [reply app/models/concerns/wranglable.rb:5]: I don't have anything to say about 2 or 4 at the moment, but if we can not count creating tags (even through the New Tag page) as wrangling activity, period, that would be fine!

File: app/views/tag_wranglers/show.html.erb
Comment (brianjaustin) [reply app/views/tag_wranglers/show.html.erb:22]: I'm thinking yes, in case they were at one point but are no longer assigned any fandoms at the moment, but I'll defer to @sarken here

Comment (sarken) [reply app/views/tag_wranglers/show.html.erb:22]: Displaying it in that case sounds fine to me.

File: app/models/concerns/wranglable.rb
Comment (sarken) [reply app/models/concerns/wranglable.rb]: Tag Wrangling says they can live with the current behavior for 2 and 4.

Comment (brianjaustin) [reply app/models/concerns/wranglable.rb]: I've added some tests that cover these cases, so I think we should be good to go with the hybrid approach. It feels a bit weird to put the flag `User`, but that felt like a better fit than some of my initial experiments with putting it on `Tag`

File: app/models/concerns/wrangleable.rb
Comment (tickinginstant) [inline app/models/concerns/wrangleable.rb:12]: I don't think this is the right place to clear `update_wrangling_activity`. If it gets set to `true`, but this hook doesn't run (e.g. if one of the validations fails in `TagsController#update`), it won't get cleared. That means it'll stay true for all subsequent requests involving the same Unicorn worker, at least until it gets cleared by a later, successful save. Possibly one by a different user that wasn't supposed to count as wrangling activity.

IMHO, it'd be better to write a controller method like this:
```ruby
def record_wrangling_activity
  User.update_wrangling_activity = true
  yield
ensure
  User.update_wrangling_activity = false
end
```
If you add that and call `around_action :record_wrangling_activity` for any controller/action combos that should be recorded, you guarantee that `User.update_wrangling_activity` will be set for any actions you want to record, and unset the rest of the time. Using `ensure` means that it'll run even if there are errors, so there's no danger of leaving `User.update_wrangling_activity` set when it's not supposed to be.

File: app/models/user.rb
Comment (tickinginstant) [inline app/models/user.rb:116]: Personally, I'd prefer `thread_cattr_accessor` here. It's not relevant for production, because each Unicorn worker is a separate process, but I believe this could cause issues if someone wanted to run a multi-threaded server instead.

File: db/migrate/20211230202951_create_last_wrangling_activities.rb
Comment (tickinginstant) [inline db/migrate/20211230202951_create_last_wrangling_activities.rb:5]: I think [the current plan is to avoid foreign keys](https://otwarchive.atlassian.net/browse/AO3-5690?focusedCommentId=358649), since they don't play nicely with online schema migrations on large tables.

File: app/models/concerns/wrangleable.rb
Comment (brianjaustin) [reply app/models/concerns/wrangleable.rb]: Yeah, that makes sense! I think I read about that kind of filter in the docs and... promptly forgot all about i :facepalm: Mòran taing!

**Review by sarken — COMMENTED** [review]
I had a few nitpicky comments saved because I wanted to come back and do a more thorough review, but I don't think I'm going to have time for that soon. So I'm going to leave the pending comments and I'll see if I can find someone willing to do a better review.

File: spec/controllers/tags_controller_spec.rb
Comment (brianjaustin) [reply spec/controllers/tags_controller_spec.rb]: For the tests with shared example [`"set last wrangling activity"`](https://github.com/otwcode/otwarchive/pull/4129/files/e45ccd7dd8750eefa0411e9de1ddb3d09f60e1d9#diff-e1135e966d22b227e95c7a9676f4ad1384cd8e1a09920aa74e9c451d0c88a6fbR11), I initially had a check for last wrangling time being equal to now. Somewhere along the line I switch to just check that timestamp is within the last minute -- I think because there was a weird precision thing somewhere -- but never removed these. I've done that now, but if you'd prefer I track the precision thing down and use `freeze_time` again I can also do that

**Review by tickinginstant — COMMENTED** [review]
Looking pretty good! Most of my comments are on style & consistency, though there is a test that doesn't seem to do what it's supposed to, and there's one earlier issue that may have been mistakenly marked as resolved.

File: spec/models/comment_spec.rb
Comment (brianjaustin) [reply spec/models/comment_spec.rb]: Oh, that `not_to change` thing is pretty cool, thanks for the tip!

File: spec/models/concerns/wrangleable_spec.rb
Comment (brianjaustin) [reply spec/models/concerns/wrangleable_spec.rb:16]: Added a cleanup to the `spec_helper`; I can move it to this one in particular but I kinda like the idea of things being cleaned up even if someone in the future also forgets

File: spec/models/comment_spec.rb
Comment (brianjaustin) [reply spec/models/comment_spec.rb:176]: Not really. I think some tests might've been flaky, hence the more generous window but I don't 100% remember. I think 1 minute should be acceptable, and if you're also fine with that I can switch the other ones from `10.seconds` to `1.minute`.

File: app/models/comment.rb
Comment (brianjaustin) [reply app/models/comment.rb]: I think I've got something working that consolidates most of that code in `User`, what do you think?

File: spec/models/concerns/wrangleable_spec.rb
Comment (tickinginstant) [reply spec/models/concerns/wrangleable_spec.rb]: Good point! Probably best to keep it in `spec_helper` then.

File: app/models/comment.rb
Comment (tickinginstant) [reply app/models/comment.rb]: Looks nice! Much DRYer.

File: app/views/tag_wranglers/show.html.erb
Comment (tickinginstant) [reply app/views/tag_wranglers/show.html.erb]: Unfortunately, it looks like the issue actually specifies that the "NAME last wrangled at DATE" line should be below the "Fandoms assigned to you via the tag wranglers page" line, so it may be necessary to either have multiple copies of the line, or check `@fandoms.empty?` more than once. I don't like the repetition, but I don't know how else to do it.

(FYI, I'm not an official volunteer, so I don't have the power to mark things as resolved in your PR.)

File: spec/models/comment_spec.rb
Comment (tickinginstant) [reply spec/models/comment_spec.rb:176]: 1 minute of leeway seems a little much, but I just had a thought: what about introducing a new conditional hook in `spec/spec_helper.rb`, like the search hooks?
https://github.com/otwcode/otwarchive/blob/99b11941fb41cefd1904a222bdf370de79e900a8/spec/spec_helper.rb#L89-L91
Those hooks are defined before all other `before` hooks, so they're also evaluated before all other `before` hooks. Which means that if you had a hook like this:
```ruby
 config.before :each, :frozen do 
   freeze_time
 end 
```
That would allow you to mark an `it` block as `:frozen` in order to prepend a call to `freeze_time` for just that example, and not break any of the existing examples that might rely on having time flow properly. Do you think that would help fix the issues you had with `freeze_time`?

File: db/migrate/20211230202951_create_last_wrangling_activities.rb
Comment (tickinginstant) [inline db/migrate/20211230202951_create_last_wrangling_activities.rb:1]: Can you update this to 6.0? 5.2 defaults to timestamps precise down to the second, whereas I think 6.0 defaults to timestamps precise down to the microsecond.

(I was going to ask you to add a `travel_to` call back to the `not_to change` tests, since those could cause misleading results if the first action and the second action occur during the same second. But I don't think you need to change those tests again if you have more precise timestamps, and since you're creating a new table, you can have timestamps as precise as you'd like.)

File: app/views/tag_wranglers/show.html.erb
Comment (brianjaustin) [reply app/views/tag_wranglers/show.html.erb]: Argh, you're right. I think I could add a partial to save some duplication, but in this case I'm not sure that would be worth it

File: spec/models/comment_spec.rb
Comment (brianjaustin) [reply spec/models/comment_spec.rb]: Fancy! It looks like that (or the migration change, or some combination of both) fixed whatever I was seeing a few months ago. Thanks for the suggestion!

File: app/views/tag_wranglers/show.html.erb
Comment (tickinginstant) [reply app/views/tag_wranglers/show.html.erb]: I think the partial might be worth it (or changing the structure of the `if` statements), because it's going to be annoying to copy it over every time it needs to be changed.

File: app/views/tag_wranglers/show.html.erb
Comment (tickinginstant) [inline app/views/tag_wranglers/show.html.erb:19]: Two extra things I just noticed:
1. In the preview screenshot, it looks like there's an extra space between the time and the period after it. That comes from `time_in_zone`, which includes an extra space at the end in some (but not all) cases. It might be good to strip the extra space, either here or in `time_in_zone`.
2. Now that you're using `t()` instead of `ts()`, it would probably be good to follow the [internationalization standards](https://github.com/otwcode/otwarchive/wiki/Internationalization-(i18n)-Standards), which specify that the whole sentence should be a single translatable string (in case there are languages where the positions of the interpolated variables in the sentence differ from English).

   In other words, it'd be good to define the string as `"%{wrangler_login} last wrangled at %{time}."`, and pass `tag.span(time_in_zone(@wrangler.last_wrangling_activity.updated_at), class: "datetime")` as an extra param.

   (To prevent the HTML from being escaped, you might need to change the key name to end in `_html`. The alternative would be manually marking the string as `html_safe`, but I'm not sure how Brakeman would feel about that.)

   Alternatively, you can switch back to `ts()` and leave it for whoever ends up internationalizing this view.

Comment (brianjaustin) [reply app/views/tag_wranglers/show.html.erb]: I ended up removing the extra space in `time_in_zone` to avoid some annoying chaining with `.strip.html_safe`. I checked the usage in templates and that seems safe, plus the attached screenshot.

existing_time_in_zone

Thanks for the tips and reminder! I've updated the template, sticking with `t()` since I've been asked to switch on similar changes in the past.

Comment (tickinginstant) [reply app/views/tag_wranglers/show.html.erb]: Did you check it with challenge dates? The one case where I'm concerned that space might be necessary is when viewing challenge dates logged in as a user whose time zone doesn't match the challenge's.

Comment (brianjaustin) [reply app/views/tag_wranglers/show.html.erb]: Right you are! I've re-shuffled things a bit to make sure that works as well, here's a screenshot after those changes.

challenge with closing date in a different time zone

Thanks for keeping me honest!

**Review by tickinginstant — APPROVED** [review]
Looks good to me! 👍

(And thanks for putting up with all of my nitpicking! 😄)

File: app/views/tag_wranglers/show.html.erb
Comment (sarken) [inline app/views/tag_wranglers/show.html.erb:21]: Could you switch this to just `logged_in_as_admin`? I mentioned to the Tag Wrangling chairs that we were thinking about putting this in the next release, but they want to give wranglers a lot of advance warning about wrangler-facing changes, so... it's either switch this to just admin-visible or wait. Just admin-visible seems like the better option.

Comment (brianjaustin) [reply app/views/tag_wranglers/show.html.erb]: Sounds reasonable to me! Is there a follow up story to add this back I can link (unfortunately I don't have permissions to create one if needed) before https://otwarchive.atlassian.net/browse/AO3-6028 or should I just update that one to include this?

Comment (sarken) [reply app/views/tag_wranglers/show.html.erb]: I checked back in to see if they wanted exposing the date handled as part of AO3-6028 or separately, and... it turns out they have now written the email. So I'm really sorry, but could you revert the last commit and return it to being visible to the relevant wrangler?

**Pull request merged**
